src/xserver.cpp

```diff
@@ -0,0 +1,77 @@
+/***************************************************************************
+* Copyright (c) 2019, Martin Renou, Johan Mabille, Sylvain Corlay and      *
+* Wolf Vollprecht                                                          *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include "xeus/xserver_zmq.hpp"
+
+#include "xserver.hpp"
+
+#include "pybind11/pybind11.h"
+
+namespace py = pybind11;
+
+namespace xpyt
+{
+    server::server(zmq::context_t& context, const xeus::xconfiguration& config)
+        : xeus::xserver_zmq(context, config)
+    {
+    }
+
+    server::~server()
+    {
+    }
+
+    void server::start_impl(zmq::multipart_t& message)
+    {
+        start_publisher_thread();
+        start_heartbeat_thread();
+
+        m_request_stop = false;
+
+        publish(message);
+
+        while (!m_request_stop)
+        {
+            poll_msg();
+        }
+
+        stop_channels();
+
+        std::exit(0);
+
+    }
+
+    void server::poll_msg()
+    {
+        zmq::pollitem_t items[]
+            = { { m_controller, 0, ZMQ_POLLIN, 0 }, { m_shell, 0, ZMQ_POLLIN, 0 } };
+
+        std::size_t nb_msg = zmq::poll(&items[0], 2, 0);
+        if (nb_msg > 0)
+        {
+            if (items[0].revents & ZMQ_POLLIN)
+            {
+                zmq::multipart_t wire_msg;
+                wire_msg.recv(m_controller);
+                xserver::notify_control_listener(wire_msg);
+            }
+
+            if (!m_request_stop && (items[1].revents & ZMQ_POLLIN))
+            {
+                zmq::multipart_t wire_msg;
+                wire_msg.recv(m_shell);
+                xserver::notify_shell_listener(wire_msg);
+            }
+        }
+        else
+        {
+            // Give some time for Python threads to resume their job
+            py::module::import("time").attr("sleep")(0.02);
```

Comment:
I'm not sure if `20ms` is fine

Reply:
`20ms` seems fine. I tested running both a Python thread on the background and a slider widget communication, even when moving the slider the thread keeps running.

Reply:
This is not relevant anymore, as we are not sleeping